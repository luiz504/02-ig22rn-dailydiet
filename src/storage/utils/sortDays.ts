export const sortDays = (array: string[]) => {
  return array.sort((a, b) => {
    return new Date(b).getTime() - new Date(a).getTime()
  })
}
