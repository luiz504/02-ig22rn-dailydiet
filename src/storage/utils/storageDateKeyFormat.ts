import { format } from 'date-fns'

export function storageDateKeyFormat(date: Date): string {
  return format(date, 'yyyy/MM/dd')
}
