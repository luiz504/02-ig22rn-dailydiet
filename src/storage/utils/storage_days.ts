import AsyncStorage from '@react-native-async-storage/async-storage'
import { DAY_COLLECTION } from '../commons'

export async function getStoredDays() {
  return AsyncStorage.getItem(DAY_COLLECTION).then((storedData) =>
    storedData ? (JSON.parse(storedData) as string[]) : null,
  )
}

export async function setStoredDays(dayCollection: string[]) {
  return AsyncStorage.setItem(DAY_COLLECTION, JSON.stringify(dayCollection))
}
