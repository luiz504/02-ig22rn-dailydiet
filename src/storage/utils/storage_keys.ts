import { format, parse } from 'date-fns'

const FORMAT = 'yyyy/MM/dd'
export function storageDateKeyFormat(date: Date): string {
  return format(date, FORMAT)
}

export function storageKeyToDateParse(key: string) {
  return parse(key, FORMAT, new Date())
}
