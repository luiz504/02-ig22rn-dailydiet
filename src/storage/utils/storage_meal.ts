import AsyncStorage from '@react-native-async-storage/async-storage'
import { MEAL_COLLECTION } from '../commons'
import { Meal } from '~/models/Meal'

export type MealGroupByDay = {
  day: string
  meals: Meal[]
}

export async function getStoredMealsByDay(key: string) {
  return AsyncStorage.getItem(`${MEAL_COLLECTION}-${key}`).then((storedData) =>
    storedData ? (JSON.parse(storedData) as MealGroupByDay) : null,
  )
}

export async function setStoredMeals(key: string, meals: Meal[]) {
  const sortedMealsByHour = meals.sort(
    (a, b) => new Date(b.date).getTime() - new Date(a.date).getTime(),
  )
  const mealsCollection: MealGroupByDay = { day: key, meals: sortedMealsByHour }

  return AsyncStorage.setItem(
    `${MEAL_COLLECTION}-${key}`,
    JSON.stringify(mealsCollection),
  )
}

export async function getStoredMealsByDays(keys: string[]) {
  const keysCombined = keys.map((k) => `${MEAL_COLLECTION}-${k}`)
  const entries = await AsyncStorage.multiGet(keysCombined)
  return entries
    .map(([, value]) => (value ? JSON.parse(value) : null))
    .filter((value) => value) as MealGroupByDay[]
}

export async function deleteStoredMealCollection(key: string) {
  return AsyncStorage.removeItem(`${MEAL_COLLECTION}-${key}`)
}
