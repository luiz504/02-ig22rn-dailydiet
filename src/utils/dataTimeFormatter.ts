import * as Localization from 'expo-localization'

const userConfig = {
  uses24hoursClock:
    Localization?.getCalendars?.()?.[0]?.uses24hourClock ?? true,
  languageTag: Localization?.getLocales?.()?.[0]?.languageTag ?? 'en-US',
}

function formatDate(date: Date) {
  return new Intl.DateTimeFormat(userConfig.languageTag, {
    year: 'numeric',
    month: '2-digit',
    day: '2-digit',
  }).format(date)
}

function formatTime(time: Date) {
  return new Intl.DateTimeFormat(userConfig.languageTag, {
    hour: '2-digit',
    minute: '2-digit',
    hour12: !userConfig.uses24hoursClock,
  }).format(time)
}

function formatDateAndTime(date: Date | string) {
  const formattedFullDate = new Date(date).toLocaleDateString(
    userConfig.languageTag,
    {
      year: 'numeric',
      month: '2-digit',
      day: '2-digit',
      hour: 'numeric',
      minute: 'numeric',
    },
  )

  if (userConfig.languageTag.startsWith('en')) {
    return formattedFullDate.replace(',', ' at')
  }

  if (userConfig.languageTag.startsWith('pt')) {
    return formattedFullDate.replace(',', ' Ã s')
  }

  return formattedFullDate
}
export { userConfig, formatDate, formatTime, formatDateAndTime }
