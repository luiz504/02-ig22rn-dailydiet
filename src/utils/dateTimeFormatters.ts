import * as Localization from 'expo-localization'

const getFirstUserConfig = () => ({
  uses24hoursClock: Localization.getCalendars()[0].uses24hourClock,
  languageTag: Localization.getLocales()[0].languageTag,
})

function formatDate(date: Date) {
  const { languageTag } = getFirstUserConfig()

  return new Intl.DateTimeFormat(languageTag, {
    year: 'numeric',
    month: '2-digit',
    day: '2-digit',
  }).format(date)
}

function formatTime(time: Date) {
  const { languageTag, uses24hoursClock } = getFirstUserConfig()

  return new Intl.DateTimeFormat(languageTag, {
    hour: '2-digit',
    minute: '2-digit',
    hour12: !uses24hoursClock,
  }).format(time)
}

function formatDateAndTime(date: Date | string) {
  const { languageTag, uses24hoursClock } = getFirstUserConfig()

  const formattedFullDate = new Date(date).toLocaleDateString(languageTag, {
    year: 'numeric',
    month: '2-digit',
    day: '2-digit',
    hour: '2-digit',
    minute: '2-digit',
    hour12: !uses24hoursClock,
  })

  if (languageTag.startsWith('en')) {
    return formattedFullDate.replace(',', ' at')
  }

  if (languageTag.startsWith('pt')) {
    return formattedFullDate.replace(',', ' Ã s')
  }

  return formattedFullDate
}

export { getFirstUserConfig, formatDate, formatTime, formatDateAndTime }
