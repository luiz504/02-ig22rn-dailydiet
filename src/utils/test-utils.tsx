import { FC, ReactNode } from 'react'
import { ThemeProvider } from 'styled-components/native'
import { render } from '@testing-library/react-native'

import { theme } from '~/styles'
import { NavigationContainer } from '@react-navigation/native'

const AllTheProviders: FC<{ children: ReactNode }> = ({ children }) => {
  return (
    <ThemeProvider theme={theme}>
      <NavigationContainer>{children}</NavigationContainer>
    </ThemeProvider>
  )
}

const ThemedWrapper: FC<{ children: ReactNode }> = ({ children }) => {
  return <ThemeProvider theme={theme}>{children}</ThemeProvider>
}

const customRender: typeof render = (ui, options) =>
  render(ui, { wrapper: AllTheProviders, ...options })

const renderThemed: typeof render = (ui, options) =>
  render(ui, { wrapper: ThemedWrapper, ...options })

const splitBorderColor = (borderColor: string) => ({
  borderTopColor: borderColor,
  borderRightColor: borderColor,
  borderBottomColor: borderColor,
  borderLeftColor: borderColor,
})

export * from '@testing-library/react-native'

export { customRender as render, renderThemed, splitBorderColor }
