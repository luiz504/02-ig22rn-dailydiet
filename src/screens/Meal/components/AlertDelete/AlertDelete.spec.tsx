import { fireEvent, render, screen } from '~/utils/test-utils'
import { AlertDelete } from '.'

describe('AlertDelete Component', () => {
  it('should render correctly', () => {
    render(<AlertDelete show onDismiss={() => {}} onConfirm={() => {}} />)

    const modalElement = screen.getByTestId('alert-delete-modal')
    const { transparent, animationType } = modalElement.props
    const overlayElement = screen.getByTestId('alert-delete-overlay')
    const content = screen.getByTestId('alert-delete-content')
    const heading = screen.getByTestId('alert-delete-heading')
    const btsRow = screen.getByTestId('alert-delete-btns-row')
    const btnDismiss = screen.getByTestId('alert-delete-btn-dismiss')
    const btnConfirm = screen.getByTestId('alert-delete-btn-confirm')

    // Assert
    expect(modalElement).toBeVisible()
    expect({ transparent, animationType }).toEqual({
      transparent: true,
      animationType: 'fade',
    })

    expect(overlayElement.props.style).toMatchSnapshot('overlay')
    expect(content.props.style).toMatchSnapshot('content')

    expect({
      style: heading.props.style,
      text: heading.props.children,
    }).toMatchSnapshot('heading')

    expect(btsRow.props.style).toMatchSnapshot('btns-row')
    expect(btnDismiss.props.style).toMatchSnapshot('btn-dismiss')
    expect(btnDismiss).toHaveTextContent('Cancel')
    expect(btnConfirm.props.style).toMatchSnapshot('btn-confirm')
    expect(btnConfirm).toHaveTextContent('Yes, delete')
  })

  it('should trigger callbacks passed by props correctly', () => {
    const onDismiss = jest.fn()
    const onConfirm = jest.fn()

    render(<AlertDelete show onDismiss={onDismiss} onConfirm={onConfirm} />)

    const btnDismiss = screen.getByTestId('alert-delete-btn-dismiss')
    const btnConfirm = screen.getByTestId('alert-delete-btn-confirm')

    // Act

    fireEvent.press(btnDismiss)

    // Asset
    expect(onDismiss).toBeCalledTimes(1)
    expect(onDismiss).toBeCalledWith()
    expect(onConfirm).toBeCalledTimes(0)

    //
    jest.clearAllMocks()

    // Act
    fireEvent.press(btnConfirm)

    expect(onDismiss).toBeCalledTimes(0)
    expect(onConfirm).toBeCalledTimes(1)
    expect(onConfirm).toBeCalledWith()
  })
})
