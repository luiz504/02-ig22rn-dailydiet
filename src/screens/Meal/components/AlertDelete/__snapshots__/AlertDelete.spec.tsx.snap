// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AlertDelete Component should render correctly: btn-confirm 1`] = `
[
  {
    "alignItems": "center",
    "borderBottomLeftRadius": 6,
    "borderBottomRightRadius": 6,
    "borderBottomWidth": 1.5,
    "borderLeftWidth": 1.5,
    "borderRightWidth": 1.5,
    "borderStyle": "solid",
    "borderTopLeftRadius": 6,
    "borderTopRightRadius": 6,
    "borderTopWidth": 1.5,
    "justifyContent": "center",
    "paddingBottom": 14,
    "paddingLeft": 24,
    "paddingRight": 24,
    "paddingTop": 14,
  },
  {
    "backgroundColor": "#333638",
    "borderColor": "#333638",
    "flex": 1,
  },
]
`;

exports[`AlertDelete Component should render correctly: btn-dismiss 1`] = `
[
  {
    "alignItems": "center",
    "borderBottomLeftRadius": 6,
    "borderBottomRightRadius": 6,
    "borderBottomWidth": 1.5,
    "borderLeftWidth": 1.5,
    "borderRightWidth": 1.5,
    "borderStyle": "solid",
    "borderTopLeftRadius": 6,
    "borderTopRightRadius": 6,
    "borderTopWidth": 1.5,
    "justifyContent": "center",
    "paddingBottom": 14,
    "paddingLeft": 24,
    "paddingRight": 24,
    "paddingTop": 14,
  },
  {
    "backgroundColor": "#FAFAFA",
    "borderColor": "#1B1D1E",
    "flex": 1,
  },
]
`;

exports[`AlertDelete Component should render correctly: btns-row 1`] = `
{
  "flexDirection": "row",
  "gap": 12,
}
`;

exports[`AlertDelete Component should render correctly: content 1`] = `
{
  "alignItems": "center",
  "alignSelf": "stretch",
  "backgroundColor": "#FAFAFA",
  "borderBottomLeftRadius": 8,
  "borderBottomRightRadius": 8,
  "borderTopLeftRadius": 8,
  "borderTopRightRadius": 8,
  "gap": 32,
  "paddingBottom": 24,
  "paddingLeft": 24,
  "paddingRight": 21,
  "paddingTop": 40,
}
`;

exports[`AlertDelete Component should render correctly: heading 1`] = `
{
  "style": {
    "color": "#1B1D1E",
    "fontFamily": "Nunito-Regular",
    "fontSize": 18,
    "lineHeight": 23.4,
  },
  "text": "Do you really want to delete the meal record?",
}
`;

exports[`AlertDelete Component should render correctly: overlay 1`] = `
{
  "alignItems": "center",
  "backgroundColor": "rgba(0,0,0,0.06)",
  "flexBasis": 0,
  "flexGrow": 1,
  "flexShrink": 1,
  "justifyContent": "center",
  "paddingBottom": 24,
  "paddingLeft": 24,
  "paddingRight": 24,
  "paddingTop": 24,
}
`;
